sql: support `corr()`

hueypark opened the pull request:
Support aggregate function `corr()`. It referred to the PostgreSQL implementation, but the implementation details follow SQL 2003. See: https://www.postgresql.org/docs/10/functions-aggregate.html

See #41274

Release note (sql change): support aggregate function `corr()`

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/44628)
<!-- Reviewable:end -->

yuzefovich commented:
Thanks for working on this!

I briefly skimmed the PR, and I think it is missing one important part - when a new aggregate function is introduced, it should be added to `pkg/sql/execinfrapb/processors_sql.proto` so that we could serialize it correctly.

hueypark commented:
Thank you for the review!

> I briefly skimmed the PR, and I think it is missing one important part - when a new aggregate function is introduced, it should be added to `pkg/sql/execinfrapb/processors_sql.proto` so that we could serialize it correctly.

I already have corr in [`pkg/sql/execinfrapb/processors_sql.proto`](https://github.com/cockroachdb/cockroach/blob/37d4c61d1ba27f3546682b6beaaef226f4e90289/pkg/sql/execinfrapb/processors_sql.proto#L471). What additional changes do we need?

yuzefovich commented:
Oops, my bad, indeed. Like I said, I briefly skimmed the PR, apparently too briefly :) I'll give a more thorough review tomorrow.

yuzefovich reviewed (commented):
Very nice work!

I have some comments and some nits.

> Reviewed 15 of 15 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44628#-:-M-BCfkbBXyy7J0J0fHR:bglnveo)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @hueypark and @RaduBerinde)

___
*[pkg/sql/opt_exec_factory.go, line 448 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44628#-M-B8SYeD-ob-e_cSjn1:-M-B8SYf1SJpKaA3iO8_:b-97oyq2) ([raw file](https://github.com/cockroachdb/cockroach/blob/37d4c61d1ba27f3546682b6beaaef226f4e90289/pkg/sql/opt_exec_factory.go#L448)):*
> ```Go
> 		agg := &aggregations[i]
> 		builtin := agg.Builtin
> 		var renderIdx []int
> ```

[nit]: `s/renderIdx/renderIdxs/g`.
___
*[pkg/sql/walk.go, line 391 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44628#-M-B8exkDW3ZAblgEu69:-M-B8exkDW3ZAblgEu6A:b-fumlkg) ([raw file](https://github.com/cockroachdb/cockroach/blob/37d4c61d1ba27f3546682b6beaaef226f4e90289/pkg/sql/walk.go#L391)):*
> ```Go
> 				} else {
> 					fmt.Fprintf(&buf, "%s(", agg.funcName)
> 					if 0 < len(agg.argRenderIdxs) {
> ```

[super nit]: it's slightly unusual to see the comparison against a number when the number is on the left-hand side.
___
*[pkg/sql/logictest/testdata/logic_test/aggregate, line 1244 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44628#-M-B91uU6nn-p2JCSN6Z:-M-B91uU6nn-p2JCSN6_:b-gekenv) ([raw file](https://github.com/cockroachdb/cockroach/blob/37d4c61d1ba27f3546682b6beaaef226f4e90289/pkg/sql/logictest/testdata/logic_test/aggregate#L1244)):*
> ```
> 
> query RR
> SELECT corr(y, x)::decimal, corr(int_y, int_x)::decimal FROM statistics_agg_test;
> ```

[nit]: a semicolon is not needed in the single statement logic test.
___
*[pkg/sql/logictest/testdata/logic_test/aggregate, line 1247 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44628#-M-BABlk4AlQ5qAX-oYU:-M-BABlk4AlQ5qAX-oYV:b-l8ctpq) ([raw file](https://github.com/cockroachdb/cockroach/blob/37d4c61d1ba27f3546682b6beaaef226f4e90289/pkg/sql/logictest/testdata/logic_test/aggregate#L1247)):*
> ```
> ----
> 0.9326733179802503 0.9326733179802503
> ```

It'd be nice to see an out of range case here as well.
___
*[pkg/sql/opt/exec/execbuilder/relational.go, line 922 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44628#-M-B9Xzc3eSUdX_Afm90:-M-B9Xzc3eSUdX_Afm91:bxxh5yc) ([raw file](https://github.com/cockroachdb/cockroach/blob/37d4c61d1ba27f3546682b6beaaef226f4e90289/pkg/sql/opt/exec/execbuilder/relational.go#L922)):*
> ```Go
> 		var filterOrd exec.ColumnOrdinal = -1
> 
> 		switch item.Agg.Op() {
> ```

I think that special casing for `CorrOp` here is not necessary. Could you refactor what is now `default` case to support multiple children?
___
*[pkg/sql/opt/memo/extract.go, line 120 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44628#-M-B9kkY48dirzZ6uto5:-M-B9kkY48dirzZ6uto6:bmy3ppp) ([raw file](https://github.com/cockroachdb/cockroach/blob/37d4c61d1ba27f3546682b6beaaef226f4e90289/pkg/sql/opt/memo/extract.go#L120)):*
> ```Go
> 	}
> 
> 	switch e.Op() {
> ```

ditto for no special casing.
___
*[pkg/sql/sem/builtins/aggregate_builtins.go, line 941 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44628#-M-BAN8U1bwvJbjnpNgu:-M-BAN8U1bwvJbjnpNgv:boibb7e) ([raw file](https://github.com/cockroachdb/cockroach/blob/37d4c61d1ba27f3546682b6beaaef226f4e90289/pkg/sql/sem/builtins/aggregate_builtins.go#L941)):*
> ```Go
> 
> // Add implements tree.AggregateFunc interface.
> func (a *corrAggregate) Add(_ context.Context, datumY tree.Datum, otherArgs ...tree.Datum) error {
> ```

Hm, the order of arguments seems suspicious to me - I'd expect `datumX` to be mentioned explicitly and `datumY` coming for the variable-args.
___
*[pkg/sql/sem/builtins/aggregate_builtins.go, line 973 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44628#-M-BCEcpA-BMkYndaXtp:-M-BCEcq2I8hl3K13_sG:bhioqtc) ([raw file](https://github.com/cockroachdb/cockroach/blob/37d4c61d1ba27f3546682b6beaaef226f4e90289/pkg/sql/sem/builtins/aggregate_builtins.go#L973)):*
> ```Go
> 		math.IsInf(a.sy2, 0) ||
> 		math.IsInf(a.sxy, 0) {
> 		return pgerror.New(pgcode.NumericValueOutOfRange, "float out of range")
> ```

[nit]: you could create this error once. Or you could export `errFloatOutOfRange` from `pkg/sql/sem/tree/eval.go` and reuse it here.
___
*[pkg/sql/sem/builtins/aggregate_builtins.go, line 1011 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44628#-M-BC_zlCy0062rdmjzf:-M-BC_zlCy0062rdmjzg:bsyj34c) ([raw file](https://github.com/cockroachdb/cockroach/blob/37d4c61d1ba27f3546682b6beaaef226f4e90289/pkg/sql/sem/builtins/aggregate_builtins.go#L1011)):*
> ```Go
> // Reset implements tree.AggregateFunc interface.
> func (a *corrAggregate) Reset(context.Context) {
> 	//a.count = 0
> ```

This seems incomplete.


<!-- Sent from Reviewable.io -->

yuzefovich commented:
The empty comment was:
> 
> 		agg := &aggregations[i]
> 		builtin := agg.Builtin
> 		var renderIdx []int
> 

(nit) `s/renderIdx/renderIdxs/g`.

Also, it might be beneficial to type out the formula for computing the correlation somewhere in `corrAggregate`.

andy-kimball reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44628)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @hueypark, @RaduBerinde, and @yuzefovich)

___
*[pkg/sql/logictest/testdata/logic_test/aggregate, line 1248 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44628#-M-Cht8Z44g7nUGyw4md:-M-Cht8Z44g7nUGyw4me:b-xvcjmo) ([raw file](https://github.com/cockroachdb/cockroach/blob/37d4c61d1ba27f3546682b6beaaef226f4e90289/pkg/sql/logictest/testdata/logic_test/aggregate#L1248)):*
> ```
> 0.9326733179802503 0.9326733179802503
> 
> subtest string_agg
> ```

I think we need a wider range of tests. Some tests that come to mind:
1. Correlation of -1, 0, 1
2. NaN
3. Out of range (as Yahor says)
4. NULL values

Think hard about what other tests would be interesting, making sure you handle all the cases in the code you wrote.
___
*[pkg/sql/opt/exec/execbuilder/relational.go, line 922 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44628#-M-B9Xzc3eSUdX_Afm90:-M-Cnu1OENvQ94WCZ0-v:b32v57) ([raw file](https://github.com/cockroachdb/cockroach/blob/37d4c61d1ba27f3546682b6beaaef226f4e90289/pkg/sql/opt/exec/execbuilder/relational.go#L922)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

I think that special casing for `CorrOp` here is not necessary. Could you refactor what is now `default` case to support multiple children?
</blockquote></details>

+1
___
*[pkg/sql/opt/exec/execbuilder/relational.go, line 928 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44628#-M-CluX8-nsTsKtYYpKx:-M-CluX94TjFzvocT7iO:b-5kc0y5) ([raw file](https://github.com/cockroachdb/cockroach/blob/37d4c61d1ba27f3546682b6beaaef226f4e90289/pkg/sql/opt/exec/execbuilder/relational.go#L928)):*
> ```Go
> 				child := item.Agg.Child(j)
> 
> 				if aggFilter, ok := child.(*memo.AggFilterExpr); ok {
> ```

`AggFilterExpr` and `AggDistinctExpr` should only occur once. We've never done multiple argument aggregates before, and I'm not sure they'll work correctly with this new function. +@raduberinde for his perspective on this.

Separately, this shows a need for having thorough tests for FILTER, DISTINCT, and ORDER BY used with the CORR aggregate function. For example:
```
select corr(distinct y, x order by x, y) filter (where x>1 and y>1) from statistics_agg_test;
```
The input values would have to be more interesting, containing duplicates and values both greater and less than 1 for `x` and `yes`. It's a tricky set of cases, and unfortunately I'm not sure we can handle these cases correctly today.
___
*[pkg/sql/opt/memo/extract.go, line 120 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44628#-M-B9kkY48dirzZ6uto5:-M-CnsbW8lVE94g72I6K:b32v57) ([raw file](https://github.com/cockroachdb/cockroach/blob/37d4c61d1ba27f3546682b6beaaef226f4e90289/pkg/sql/opt/memo/extract.go#L120)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

ditto for no special casing.
</blockquote></details>

+1


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (approved):
I also noticed that there is some optimizer builder code which assumes all but the first argument is a constant. We'll need to fix that too: https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/opt/optbuilder/groupby.go#L358

@andy-kimball and myself are considering moving the `AggDistinct` and `AggFilter` modifiers up on top of the aggregate function (similar to how windowing modifiers work). They are currently under the function, on the input side, which makes it hard to extend things to multiple arguments. I think once we do that, it will be much easier to add `corr()`.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44628#-:-M-FudCz3htVxXi9WJc8:b-vo3me4)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @hueypark, @RaduBerinde, and @yuzefovich)



<!-- Sent from Reviewable.io -->

hueypark reviewed (commented):
Thank you for the review!

> I also noticed that there is some optimizer builder code which assumes all but the first argument is a constant. We'll need to fix that too: https://github.com/cockroachdb/cockroach/blob/master/pkg/sql/opt/optbuilder/groupby.go#L358

Done.

> @andy-kimball and myself are considering moving the AggDistinct and AggFilter modifiers up on top of the aggregate function (similar to how windowing modifiers work). They are currently under the function, on the input side, which makes it hard to extend things to multiple arguments. I think once we do that, it will be much easier to add corr().

Done.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44628#-:-M-BvqKS2ZBUw4YUbfPJ:b-de5cvm)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball, @RaduBerinde, and @yuzefovich)

___
*[pkg/sql/opt_exec_factory.go, line 448 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44628#-M-B8SYeD-ob-e_cSjn1:-M-C8Jq1EA4MP8N62nW5:b-6z5x6i) ([raw file](https://github.com/cockroachdb/cockroach/blob/37d4c61d1ba27f3546682b6beaaef226f4e90289/pkg/sql/opt_exec_factory.go#L448)):*
> (nit) s/renderIdx/renderIdxs/g.
Done.

> Also, it might be beneficial to type out the formula for computing the correlation somewhere in corrAggregate.
I don't know exactly what you mean.
___
*[pkg/sql/walk.go, line 391 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44628#-M-B8exkDW3ZAblgEu69:-M-C8_1l-G6aC1X7HjIN:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/37d4c61d1ba27f3546682b6beaaef226f4e90289/pkg/sql/walk.go#L391)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

[super nit]: it's slightly unusual to see the comparison against a number when the number is on the left-hand side.
</blockquote></details>

Done.
___
*[pkg/sql/logictest/testdata/logic_test/aggregate, line 1244 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44628#-M-B91uU6nn-p2JCSN6Z:-M-C8drgBfx4HcLarfG3:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/37d4c61d1ba27f3546682b6beaaef226f4e90289/pkg/sql/logictest/testdata/logic_test/aggregate#L1244)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

[nit]: a semicolon is not needed in the single statement logic test.
</blockquote></details>

Done.
___
*[pkg/sql/logictest/testdata/logic_test/aggregate, line 1247 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44628#-M-BABlk4AlQ5qAX-oYU:-M-CBRos3_V7XocuqBVr:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/37d4c61d1ba27f3546682b6beaaef226f4e90289/pkg/sql/logictest/testdata/logic_test/aggregate#L1247)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

It'd be nice to see an out of range case here as well.
</blockquote></details>

Done.
___
*[pkg/sql/logictest/testdata/logic_test/aggregate, line 1248 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44628#-M-Cht8Z44g7nUGyw4md:-M-c10nwE0cDw4u4QDjG:b-wyr6pt) ([raw file](https://github.com/cockroachdb/cockroach/blob/37d4c61d1ba27f3546682b6beaaef226f4e90289/pkg/sql/logictest/testdata/logic_test/aggregate#L1248)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

I think we need a wider range of tests. Some tests that come to mind:
1. Correlation of -1, 0, 1
2. NaN
3. Out of range (as Yahor says)
4. NULL values

Think hard about what other tests would be interesting, making sure you handle all the cases in the code you wrote.
</blockquote></details>

Thank you for your comment on tests. Done.
___
*[pkg/sql/opt/exec/execbuilder/relational.go, line 922 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44628#-M-B9Xzc3eSUdX_Afm90:-M-H03xQ387h4mGmp4OJ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/37d4c61d1ba27f3546682b6beaaef226f4e90289/pkg/sql/opt/exec/execbuilder/relational.go#L922)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

+1
</blockquote></details>

Done.
___
*[pkg/sql/opt/exec/execbuilder/relational.go, line 928 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44628#-M-CluX8-nsTsKtYYpKx:-M-c1JMYADFyuKDesrGP:b-kfruk8) ([raw file](https://github.com/cockroachdb/cockroach/blob/37d4c61d1ba27f3546682b6beaaef226f4e90289/pkg/sql/opt/exec/execbuilder/relational.go#L928)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

`AggFilterExpr` and `AggDistinctExpr` should only occur once. We've never done multiple argument aggregates before, and I'm not sure they'll work correctly with this new function. +@raduberinde for his perspective on this.

Separately, this shows a need for having thorough tests for FILTER, DISTINCT, and ORDER BY used with the CORR aggregate function. For example:
```
select corr(distinct y, x order by x, y) filter (where x>1 and y>1) from statistics_agg_test;
```
The input values would have to be more interesting, containing duplicates and values both greater and less than 1 for `x` and `yes`. It's a tricky set of cases, and unfortunately I'm not sure we can handle these cases correctly today.
</blockquote></details>

Partly done except aggregate with both `DISTINCT` and `ORDER BY`.

Maybe we can handle [sql: aggregate with both DISTINCT and ORDER BY not supported](https://github.com/cockroachdb/cockroach/issues/43237) later.
___
*[pkg/sql/opt/memo/extract.go, line 120 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44628#-M-B9kkY48dirzZ6uto5:-M-H4WI14djXyHIcNSBq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/37d4c61d1ba27f3546682b6beaaef226f4e90289/pkg/sql/opt/memo/extract.go#L120)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

+1
</blockquote></details>

Done.
___
*[pkg/sql/sem/builtins/aggregate_builtins.go, line 941 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44628#-M-BAN8U1bwvJbjnpNgu:-M-H5Rxg4es0H66OcHL-:bbtp8rn) ([raw file](https://github.com/cockroachdb/cockroach/blob/37d4c61d1ba27f3546682b6beaaef226f4e90289/pkg/sql/sem/builtins/aggregate_builtins.go#L941)):*
> Hm, the order of arguments seems suspicious to me - I'd expect datumX to be mentioned explicitly and datumY coming for the variable-args.

At first, I was suspicious too. But I follow [PostgreSQL](https://www.postgresql.org/docs/current/functions-aggregate.html) and SQL 2003 rule.
___
*[pkg/sql/sem/builtins/aggregate_builtins.go, line 973 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44628#-M-BCEcpA-BMkYndaXtp:-M-H5sgd5fkQE5LW2kd_:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/37d4c61d1ba27f3546682b6beaaef226f4e90289/pkg/sql/sem/builtins/aggregate_builtins.go#L973)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

[nit]: you could create this error once. Or you could export `errFloatOutOfRange` from `pkg/sql/sem/tree/eval.go` and reuse it here.
</blockquote></details>

Done.
___
*[pkg/sql/sem/builtins/aggregate_builtins.go, line 1011 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44628#-M-BC_zlCy0062rdmjzf:-M-H75Ix5zcIvXu-LHIU:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/37d4c61d1ba27f3546682b6beaaef226f4e90289/pkg/sql/sem/builtins/aggregate_builtins.go#L1011)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

This seems incomplete.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

andy-kimball commented:
I just merged https://github.com/cockroachdb/cockroach/pull/44839, which prepares the optimizer for multi-argument aggregation functions.

andy-kimball commented:
One idea (but not required) is to split this work into two commits (or PR's):
- The first would be to add support for multi-argument aggregate functions by enabling `string_agg` to have 2 variable arguments (today, the 2nd arg must be constant). This PR would include a big chunk of what you've done in this PR.
- The second would be to add the `corr()` function using the framework you enabled in the first PR. This would contain the remaining chunk of work that is specific to `corr`.

The nice thing about this approach is that the 2nd PR would then become the "template" that others could follow when adding other aggregate functions in the future. As it is, this PR is really doing 2 things at once, and it can be challenging to separate what's needed for the 1st vs. what's needed for the 2nd.

hueypark reviewed (commented):
Thank you for merge!

I splitted this work into two commits.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44628#-:-M-dTPjwEsMv5NCabfND:b26yk2y)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball, @RaduBerinde, and @yuzefovich)



<!-- Sent from Reviewable.io -->

hueypark reviewed (commented):
Hmm... some TeamCity build failed for no reason. Can anyone check this?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44628#-:-M-j9sbC9SmwZvnPsHsI:b-w4np0f)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball, @RaduBerinde, and @yuzefovich)



<!-- Sent from Reviewable.io -->

RaduBerinde commented:
I restarted it. I believe we have some known flaky roachtests.

hueypark reviewed (commented):
Thank you for checking.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44628#-:-M-kvyTW7fpRTtJamXnA:btdo30h)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball, @RaduBerinde, and @yuzefovich)



<!-- Sent from Reviewable.io -->

hueypark reviewed (commented):
Can I ask when we are going to merge this?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44628#-:-M-tQiy1BvRtJ8ZLVi0d:btzv7l1)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball, @RaduBerinde, and @yuzefovich)



<!-- Sent from Reviewable.io -->

andy-kimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44628)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball, @hueypark, @RaduBerinde, and @yuzefovich)

___
*[pkg/sql/opt_exec_factory.go, line 450 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/44628#-M-tYsIN1zvarOToxwIK:-M-tYsIN1zvarOToxwIL:b-y6eybr) ([raw file](https://github.com/cockroachdb/cockroach/blob/dbcc32e0d0c50b759c697262e992d194c309ba44/pkg/sql/opt_exec_factory.go#L450)):*
> ```Go
> 		var aggFn func(*tree.EvalContext, tree.Datums) tree.AggregateFunc
> 
> 		switch len(agg.ArgCols) {
> ```

Can you generalize this into a loop, rather than having one case per #ArgCols?
___
*[pkg/sql/walk.go, line 425 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44628#-M-takApC7jEjKm4uum1:-M-takApC7jEjKm4uum2:b-difbnt) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b5940af336f3491d7a9957a2a373809937d353b/pkg/sql/walk.go#L425)):*
> ```Go
> 						}
> 						for _, idx := range agg.argRenderIdxs {
> 							buf.WriteString(inputCols[idx].Name)
> ```

You need to separate arguments with a comma, otherwise you get this:
```
root@:26257/defaultdb> explain select corr(x, y) from (values (1, 1), (2, 2)) t(x, y);
     tree     |    field    |     description
--------------+-------------+-----------------------
              | distributed | true
              | vectorized  | false
  group       |             |
   │          | aggregate 0 | corr(column1column2)
   │          | scalar      |
   └── values |             |
              | size        | 2 columns, 2 rows
```
Notice how column1 and column2 are smashed together. Also, make sure to add a test case for this in `execbuilder/testdata/explain` to make sure we handle EXPLAIN with multiple args.
___
*[pkg/sql/opt/operator.go, line 256 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/44628#-M-tc7uI6dvKwntJapXm:-M-tc7uI6dvKwntJapXn:bjmvg6b) ([raw file](https://github.com/cockroachdb/cockroach/blob/dbcc32e0d0c50b759c697262e992d194c309ba44/pkg/sql/opt/operator.go#L256)):*
> ```Go
> 
> // AggregateIgnoresNulls returns true if the given aggregate operator has a
> // single input, and if it always evaluates to the same result regardless of
> ```

`CorrOp` does not belong here, since it does not have a single input.
___
*[pkg/sql/opt/operator.go, line 269 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/44628#-M-tcCT9AmQ6o7Nb-5go:-M-tcCT9AmQ6o7Nb-5gp:bs4q9op) ([raw file](https://github.com/cockroachdb/cockroach/blob/dbcc32e0d0c50b759c697262e992d194c309ba44/pkg/sql/opt/operator.go#L269)):*
> ```Go
> 
> // AggregateIsNullOnEmpty returns true if the given aggregate operator has a
> // single input, and if it returns NULL when the input set contains no values.
> ```

Ditto.
___
*[pkg/sql/opt/optbuilder/groupby.go, line 354 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44628#-M-tdpY6CFLWoJ0yvt3o:-M-tdpY6CFLWoJ0yvt3p:b-ywicma) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b5940af336f3491d7a9957a2a373809937d353b/pkg/sql/opt/optbuilder/groupby.go#L354)):*
> ```Go
> 		args := make([]opt.ScalarExpr, 0, 2)
> 		for i, arg := range agg.args {
> 			switch agg.def.Name {
> ```

What I had in mind here when I put the comment was to get rid of the special case for `string_agg` altogether. If you decide you want to do that, then it'd mean adding new `string_agg` related test cases in various places. If you'd rather let me do it, then leave in the comment as a marker, so I don't forget to fix `string_agg`.

Regardless of which way you go, we do need test cases for `DISTINCT` and `FILTER`, when used with multiple agg args. Those could be added in this commit, if you enable `string_agg` with multiple variable args, or they could go in the `corr` commit. We'd want those tests at multiple levels:
```
  opt/optbuilder/testdata/aggregate (for building the opt tree)
  opt/execbuilder/testdata/aggregate (for EXPLAIN case)
  opt/norm/testdata/rules/agg (for making sure that DISTINCT/FILTER rules work properly with aggs of multiple args)
  pkg/sql/logictest/testdata/logic_test/aggregate (to make sure multiple args work in regular execution)
  pkg/sql/logictest/testdata/logic_test/distsql_agg (to make sure multiple args work with distsql)
  
```

This is not easy, I know, but this is a pretty important change, and we want to make sure it's tested in depth, as it affects so many components in the query stack.

When I suggested breaking the PR into two commits, I had in mind using `string_agg` test cases as the way to test this commit. Once you've thoroughly tested multi-arg aggs by testing `string_agg`, you wouldn't need to do that same level of testing with `corr`. There, your focus would be on testing the implementation of `corr` rather than on the fact that it had multiple args (which should already be tested sufficiently by `string_agg` tests).
___
*[pkg/sql/opt/optbuilder/groupby.go, line 773 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/44628#-M-tcGHr2zhdIPCrN76F:-M-tcGHr2zhdIPCrN76G:b4vp0c7) ([raw file](https://github.com/cockroachdb/cockroach/blob/dbcc32e0d0c50b759c697262e992d194c309ba44/pkg/sql/opt/optbuilder/groupby.go#L773)):*
> ```Go
> 	case "concat_agg":
> 		return b.factory.ConstructConcatAgg(args[0])
> 	case "corr":
> ```

Anytime you change an `optbuilder` file, there should be a test for that functionality in the `testdata` directory. In this case, I'd expect a test of some kind in the `aggregate` file. In general, we try to unit test thoroughly at each layer.
___
*[pkg/sql/sem/builtins/aggregate_builtins.go, line 930 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/44628#-M-tbaq09WNpmqI93eol:-M-tbaq13Z4AzIj4lDNw:b-umqiv0) ([raw file](https://github.com/cockroachdb/cockroach/blob/dbcc32e0d0c50b759c697262e992d194c309ba44/pkg/sql/sem/builtins/aggregate_builtins.go#L930)):*
> ```Go
> }
> 
> // corrAggregate represents SQL:2003 correlation coefficient.
> ```

Add to this comment, giving a description of the correlation coefficient, perhaps copied straight from the SQL 2003 spec.


<!-- Sent from Reviewable.io -->

andy-kimball reviewed (commented):
Sorry for the delay, a lot going on. Thanks for splitting the PR into two commits. There are still various issues to address, but this is coming along. Thanks for working on it!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44628#-:-M-thN032rgcW_hFURla:b-l058vo)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball, @hueypark, @RaduBerinde, and @yuzefovich)



<!-- Sent from Reviewable.io -->

hueypark reviewed (commented):
Thank you for your review.  I'll check the details this weekend.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44628#-:-M-ttkb143xFYJUZ0syK:b3eo0e6)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball, @hueypark, @RaduBerinde, and @yuzefovich)



<!-- Sent from Reviewable.io -->

hueypark reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44628)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball, @RaduBerinde, and @yuzefovich)

___
*[pkg/sql/opt_exec_factory.go, line 450 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/44628#-M-tYsIN1zvarOToxwIK:-M-vOsqe6YDm3k76zfZX:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/dbcc32e0d0c50b759c697262e992d194c309ba44/pkg/sql/opt_exec_factory.go#L450)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Can you generalize this into a loop, rather than having one case per #ArgCols?
</blockquote></details>

Done.
___
*[pkg/sql/walk.go, line 425 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44628#-M-takApC7jEjKm4uum1:-M06tLFC45HJfjhHkdr6:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b5940af336f3491d7a9957a2a373809937d353b/pkg/sql/walk.go#L425)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

You need to separate arguments with a comma, otherwise you get this:
```
root@:26257/defaultdb> explain select corr(x, y) from (values (1, 1), (2, 2)) t(x, y);
     tree     |    field    |     description
--------------+-------------+-----------------------
              | distributed | true
              | vectorized  | false
  group       |             |
   │          | aggregate 0 | corr(column1column2)
   │          | scalar      |
   └── values |             |
              | size        | 2 columns, 2 rows
```
Notice how column1 and column2 are smashed together. Also, make sure to add a test case for this in `execbuilder/testdata/explain` to make sure we handle EXPLAIN with multiple args.
</blockquote></details>

Done.
___
*[pkg/sql/opt/operator.go, line 256 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/44628#-M-tc7uI6dvKwntJapXm:-M06tPoZD4-fZ7cuBeh9:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/dbcc32e0d0c50b759c697262e992d194c309ba44/pkg/sql/opt/operator.go#L256)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

`CorrOp` does not belong here, since it does not have a single input.
</blockquote></details>

Done.
___
*[pkg/sql/opt/operator.go, line 269 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/44628#-M-tcCT9AmQ6o7Nb-5go:-M06tQ6R8A6nRzzYgKrd:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/dbcc32e0d0c50b759c697262e992d194c309ba44/pkg/sql/opt/operator.go#L269)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Ditto.
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/groupby.go, line 354 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44628#-M-tdpY6CFLWoJ0yvt3o:-M0C3Tfj90ZkSWnFW_fu:bvlwssw) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b5940af336f3491d7a9957a2a373809937d353b/pkg/sql/opt/optbuilder/groupby.go#L354)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

What I had in mind here when I put the comment was to get rid of the special case for `string_agg` altogether. If you decide you want to do that, then it'd mean adding new `string_agg` related test cases in various places. If you'd rather let me do it, then leave in the comment as a marker, so I don't forget to fix `string_agg`.

Regardless of which way you go, we do need test cases for `DISTINCT` and `FILTER`, when used with multiple agg args. Those could be added in this commit, if you enable `string_agg` with multiple variable args, or they could go in the `corr` commit. We'd want those tests at multiple levels:
```
  opt/optbuilder/testdata/aggregate (for building the opt tree)
  opt/execbuilder/testdata/aggregate (for EXPLAIN case)
  opt/norm/testdata/rules/agg (for making sure that DISTINCT/FILTER rules work properly with aggs of multiple args)
  pkg/sql/logictest/testdata/logic_test/aggregate (to make sure multiple args work in regular execution)
  pkg/sql/logictest/testdata/logic_test/distsql_agg (to make sure multiple args work with distsql)
  
```

This is not easy, I know, but this is a pretty important change, and we want to make sure it's tested in depth, as it affects so many components in the query stack.

When I suggested breaking the PR into two commits, I had in mind using `string_agg` test cases as the way to test this commit. Once you've thoroughly tested multi-arg aggs by testing `string_agg`, you wouldn't need to do that same level of testing with `corr`. There, your focus would be on testing the implementation of `corr` rather than on the fact that it had multiple args (which should already be tested sufficiently by `string_agg` tests).
</blockquote></details>

Done.
If we want more tests. Please let me know.
___
*[pkg/sql/opt/optbuilder/groupby.go, line 773 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/44628#-M-tcGHr2zhdIPCrN76F:-M06wKZb13uS3ppYXVUr:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/dbcc32e0d0c50b759c697262e992d194c309ba44/pkg/sql/opt/optbuilder/groupby.go#L773)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Anytime you change an `optbuilder` file, there should be a test for that functionality in the `testdata` directory. In this case, I'd expect a test of some kind in the `aggregate` file. In general, we try to unit test thoroughly at each layer.
</blockquote></details>

Done.
___
*[pkg/sql/sem/builtins/aggregate_builtins.go, line 930 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/44628#-M-tbaq09WNpmqI93eol:-M06p3UN5d8uNj6yzUKh:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/dbcc32e0d0c50b759c697262e992d194c309ba44/pkg/sql/sem/builtins/aggregate_builtins.go#L930)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Add to this comment, giving a description of the correlation coefficient, perhaps copied straight from the SQL 2003 spec.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

andy-kimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44628)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball, @hueypark, @RaduBerinde, and @yuzefovich)

___
*[pkg/sql/opt_exec_factory.go, line 447 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/44628#-M0InomP9vFCBSJUjttr:-M0InomP9vFCBSJUjtts:bmbgqsq) ([raw file](https://github.com/cockroachdb/cockroach/blob/b296e2208701f486017e7a3dc5475378a69e1b53/pkg/sql/opt_exec_factory.go#L447)):*
> ```Go
> 		agg := &aggregations[i]
> 		builtin := agg.Builtin
> 		var renderIdxs []int
> ```

Should preallocate these to avoid unnecessary allocations during resizing:
```
renderIdxs := make([]int, len(agg.ArgCols));
params := make([]*types.T, len(agg.ArgCols));
```
___
*[pkg/sql/walk.go, line 429 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/44628#-M0IoKn7-SWKNVTmTWNu:-M0IoKn7-SWKNVTmTWNv:b-ncjkt5) ([raw file](https://github.com/cockroachdb/cockroach/blob/b296e2208701f486017e7a3dc5475378a69e1b53/pkg/sql/walk.go#L429)):*
> ```Go
> 							names = append(names, inputCols[idx].Name)
> 						}
> 						buf.WriteString(strings.Join(names, ", "))
> ```

NIT: Multiple unnecessary allocations here. Instead, always follow this pattern in cases like this:
```
for i, idx := range agg.argRenderIdxs {
  if i != 0 {
    buf.WriteString(", ")
  }
  buf.WriteString(inputCols[idx].Name)
}
```
___
*[pkg/sql/opt/operator.go, line 256 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/44628#-M-tc7uI6dvKwntJapXm:-M0JJejL-EKO5HGzH-aO:b-hu3cm4) ([raw file](https://github.com/cockroachdb/cockroach/blob/dbcc32e0d0c50b759c697262e992d194c309ba44/pkg/sql/opt/operator.go#L256)):*
<details><summary><i>Previously, hueypark (Jaewan Park) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

After seeing the change you had to make to `decorrelate_test.go`, I looked into this more deeply. We should change this comment to:
```
// AggregateIgnoreNulls returns true if the given aggregate operator ignores
// rows where its first argument evaluates to NULL. In other words, it always
// evaluates to the same result even if those rows are filtered. For example:
//
//   SELECT string_agg(x, y)
//   FROM (VALUES ('foo', ','), ('bar', ','), (NULL, ',')) t(x, y)
//
// In this example, the NULL row can be removed from the input, and the
// string_agg function still returns the same result. Contrast this to the
// array_agg function:
//
//   SELECT array_agg(x)
//   FROM (VALUES ('foo'), (NULL), ('bar')) t(x)
//
// If the NULL row is removed here, array_agg returns {foo,bar} instead of
// {foo,NULL,bar}.
```

Then you can add `CorrOp` back into the list, since it does seem that it skips input rows where its first argument is NULL. You should also add a couple tests. Add this to `opt/norm/testdata/rules/decorrelate`:
```
# With a multi-argument aggregate.
norm expect=TryDecorrelateScalarGroupBy
SELECT k, (SELECT string_agg(a.s, ',') FROM a WHERE a.k = a2.i) FROM a AS a2
----
```
The expected result is that the correlated subquery is turned into a regular LEFT JOIN.

And this test to `opt/norm/testdata/rules/reject_nulls`:
```
# Use with multi-argument aggregate function.
norm expect=RejectNullsGroupBy
SELECT string_agg(s, ',')
FROM (SELECT x FROM xy)
LEFT JOIN (SELECT k, s FROM a)
ON True
GROUP BY k
HAVING string_agg(s, ',')='foo'
----
```
The expected result is the LEFT JOIN gets converted into an INNER JOIN.
___
*[pkg/sql/opt/operator.go, line 269 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/44628#-M-tcCT9AmQ6o7Nb-5go:-M0JO9L13_zYCo6RlVlb:b-eljwpz) ([raw file](https://github.com/cockroachdb/cockroach/blob/dbcc32e0d0c50b759c697262e992d194c309ba44/pkg/sql/opt/operator.go#L269)):*
<details><summary><i>Previously, hueypark (Jaewan Park) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Change this comment to:
```
// AggregateIsNullOnEmpty returns true if the given aggregate operator returns
// NULL when the input set contains no values. This group of aggregates overlaps
// considerably with the AggregateIgnoresNulls group, with the notable exception
// of COUNT, which returns zero instead of NULL when its input is empty.
```
And then add back `CorrOp`.
___
*[pkg/sql/opt/norm/decorrelate_test.go, line 27 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/44628#-M0JYriqDHIfF3ZeBGmK:-M0JYriqDHIfF3ZeBGmL:budl48b) ([raw file](https://github.com/cockroachdb/cockroach/blob/b296e2208701f486017e7a3dc5475378a69e1b53/pkg/sql/opt/norm/decorrelate_test.go#L27)):*
> ```Go
> 	for op := range opt.AggregateOpReverseMap {
> 		switch op {
> 		case opt.CorrOp, opt.CountRowsOp:
> ```

Revert this once the Aggregate boolean functions in `operator.go` are updated.
___
*[pkg/sql/opt/optbuilder/groupby.go, line 354 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44628#-M-tdpY6CFLWoJ0yvt3o:-M0IqabV06q7nBgies_H:b-is7qjx) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b5940af336f3491d7a9957a2a373809937d353b/pkg/sql/opt/optbuilder/groupby.go#L354)):*
<details><summary><i>Previously, hueypark (Jaewan Park) wrote…</i></summary><blockquote>

Done.
If we want more tests. Please let me know.
</blockquote></details>

This should just be:
```
  colID := argCols[0].id
  args = append(args, b.factory.ConstructVariable(colID))
```
We just want to make *every* argument a Variable. I think there's a subtle bug with the way you're doing it for cases like:
```
string_agg(s, 'foo', s, 'bar')
```
We have an assumption when building the execution operators that all constant operands are at the end of the argument list, but that'd be violated by this code.

After fixing this, make sure to add a couple interesting test cases (in `logic_test/aggregate`).
___
*[pkg/sql/sem/builtins/aggregate_builtins.go, line 930 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/44628#-M-tbaq09WNpmqI93eol:-M0IrcfyEQAZ8kozf_Z-:bwysxc) ([raw file](https://github.com/cockroachdb/cockroach/blob/dbcc32e0d0c50b759c697262e992d194c309ba44/pkg/sql/sem/builtins/aggregate_builtins.go#L930)):*
<details><summary><i>Previously, hueypark (Jaewan Park) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Very nice comment, thanks.


<!-- Sent from Reviewable.io -->

hueypark reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44628)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball, @RaduBerinde, and @yuzefovich)

___
*[pkg/sql/opt_exec_factory.go, line 447 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/44628#-M0InomP9vFCBSJUjttr:-M0NMEYF7651hJiGcAK2:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b296e2208701f486017e7a3dc5475378a69e1b53/pkg/sql/opt_exec_factory.go#L447)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Should preallocate these to avoid unnecessary allocations during resizing:
```
renderIdxs := make([]int, len(agg.ArgCols));
params := make([]*types.T, len(agg.ArgCols));
```
</blockquote></details>

Done.
___
*[pkg/sql/walk.go, line 429 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/44628#-M0IoKn7-SWKNVTmTWNu:-M0NMSMC5mi0Q6YFg3ag:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b296e2208701f486017e7a3dc5475378a69e1b53/pkg/sql/walk.go#L429)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

NIT: Multiple unnecessary allocations here. Instead, always follow this pattern in cases like this:
```
for i, idx := range agg.argRenderIdxs {
  if i != 0 {
    buf.WriteString(", ")
  }
  buf.WriteString(inputCols[idx].Name)
}
```
</blockquote></details>

Done.
___
*[pkg/sql/opt/operator.go, line 256 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/44628#-M-tc7uI6dvKwntJapXm:-M0OAe6V0O-WL2uY1vNg:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/dbcc32e0d0c50b759c697262e992d194c309ba44/pkg/sql/opt/operator.go#L256)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

After seeing the change you had to make to `decorrelate_test.go`, I looked into this more deeply. We should change this comment to:
```
// AggregateIgnoreNulls returns true if the given aggregate operator ignores
// rows where its first argument evaluates to NULL. In other words, it always
// evaluates to the same result even if those rows are filtered. For example:
//
//   SELECT string_agg(x, y)
//   FROM (VALUES ('foo', ','), ('bar', ','), (NULL, ',')) t(x, y)
//
// In this example, the NULL row can be removed from the input, and the
// string_agg function still returns the same result. Contrast this to the
// array_agg function:
//
//   SELECT array_agg(x)
//   FROM (VALUES ('foo'), (NULL), ('bar')) t(x)
//
// If the NULL row is removed here, array_agg returns {foo,bar} instead of
// {foo,NULL,bar}.
```

Then you can add `CorrOp` back into the list, since it does seem that it skips input rows where its first argument is NULL. You should also add a couple tests. Add this to `opt/norm/testdata/rules/decorrelate`:
```
# With a multi-argument aggregate.
norm expect=TryDecorrelateScalarGroupBy
SELECT k, (SELECT string_agg(a.s, ',') FROM a WHERE a.k = a2.i) FROM a AS a2
----
```
The expected result is that the correlated subquery is turned into a regular LEFT JOIN.

And this test to `opt/norm/testdata/rules/reject_nulls`:
```
# Use with multi-argument aggregate function.
norm expect=RejectNullsGroupBy
SELECT string_agg(s, ',')
FROM (SELECT x FROM xy)
LEFT JOIN (SELECT k, s FROM a)
ON True
GROUP BY k
HAVING string_agg(s, ',')='foo'
----
```
The expected result is the LEFT JOIN gets converted into an INNER JOIN.

</blockquote></details>

Done.
___
*[pkg/sql/opt/operator.go, line 269 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/44628#-M-tcCT9AmQ6o7Nb-5go:-M0OBeao-iNSCtqsUaqN:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/dbcc32e0d0c50b759c697262e992d194c309ba44/pkg/sql/opt/operator.go#L269)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Change this comment to:
```
// AggregateIsNullOnEmpty returns true if the given aggregate operator returns
// NULL when the input set contains no values. This group of aggregates overlaps
// considerably with the AggregateIgnoresNulls group, with the notable exception
// of COUNT, which returns zero instead of NULL when its input is empty.
```
And then add back `CorrOp`.
</blockquote></details>

Done.
___
*[pkg/sql/opt/norm/decorrelate_test.go, line 27 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/44628#-M0JYriqDHIfF3ZeBGmK:-M0NXWX8BWQiHIhFYcFe:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/b296e2208701f486017e7a3dc5475378a69e1b53/pkg/sql/opt/norm/decorrelate_test.go#L27)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

Revert this once the Aggregate boolean functions in `operator.go` are updated.
</blockquote></details>

Done.
___
*[pkg/sql/opt/optbuilder/groupby.go, line 354 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44628#-M-tdpY6CFLWoJ0yvt3o:-M0O9-r23tkQpD3lbIgP:b-u1pk78) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b5940af336f3491d7a9957a2a373809937d353b/pkg/sql/opt/optbuilder/groupby.go#L354)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

This should just be:
```
  colID := argCols[0].id
  args = append(args, b.factory.ConstructVariable(colID))
```
We just want to make *every* argument a Variable. I think there's a subtle bug with the way you're doing it for cases like:
```
string_agg(s, 'foo', s, 'bar')
```
We have an assumption when building the execution operators that all constant operands are at the end of the argument list, but that'd be violated by this code.

After fixing this, make sure to add a couple interesting test cases (in `logic_test/aggregate`).
</blockquote></details>

We should be able to treat the argument as a constant. If not, we can not support `RejectNullsGroupBy` for `string_agg`.

Right now we can't treat the first argument as a constant, so I left a comment.


<!-- Sent from Reviewable.io -->

andy-kimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44628)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball, @hueypark, @RaduBerinde, and @yuzefovich)

___
*[pkg/sql/walk.go, line 425 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44628#-M-takApC7jEjKm4uum1:-M0Oi0Og7tIwIfEQVm8t:b-6vawkf) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b5940af336f3491d7a9957a2a373809937d353b/pkg/sql/walk.go#L425)):*
<details><summary><i>Previously, hueypark (Jaewan Park) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

This needs a test in `exec/execbuilder/testdata/explain`.
___
*[pkg/sql/opt/optbuilder/groupby.go, line 354 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44628#-M-tdpY6CFLWoJ0yvt3o:-M0Oij_x0okd_JCO503L:b-cxe5cw) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b5940af336f3491d7a9957a2a373809937d353b/pkg/sql/opt/optbuilder/groupby.go#L354)):*
<details><summary><i>Previously, hueypark (Jaewan Park) wrote…</i></summary><blockquote>

We should be able to treat the argument as a constant. If not, we can not support `RejectNullsGroupBy` for `string_agg`.

Right now we can't treat the first argument as a constant, so I left a comment.

</blockquote></details>

I'm not sure I understand. Can you show me the problematic case that you're worried about? I believe this should just work:
```
for i, arg := range agg.args {
  colID := argCols[0].id
  args = append(args, b.factory.ConstructVariable(colID))

  // Skip past argCols that have been handled. There may be variable
  // number of them, so need to set up for next aggregate function.
  argCols = argCols[1:]
}
```
If you look at the plans that are generated, even constants are just projected input variables to the aggregates. Once we do this, we could theoretically remove support for aggregation argument constants during execution.


<!-- Sent from Reviewable.io -->

hueypark reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44628)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball, @RaduBerinde, and @yuzefovich)

___
*[pkg/sql/walk.go, line 425 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44628#-M-takApC7jEjKm4uum1:-M0SUlxNEoOn-5dvI32Q:b-sty3x4) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b5940af336f3491d7a9957a2a373809937d353b/pkg/sql/walk.go#L425)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

This needs a test in `exec/execbuilder/testdata/explain`.
</blockquote></details>

Oh, I forgot this.
I added it now.
___
*[pkg/sql/opt/optbuilder/groupby.go, line 354 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44628#-M-tdpY6CFLWoJ0yvt3o:-M0SeM2M1aQpLNIsN3UZ:biqmfw0) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b5940af336f3491d7a9957a2a373809937d353b/pkg/sql/opt/optbuilder/groupby.go#L354)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

I'm not sure I understand. Can you show me the problematic case that you're worried about? I believe this should just work:
```
for i, arg := range agg.args {
  colID := argCols[0].id
  args = append(args, b.factory.ConstructVariable(colID))

  // Skip past argCols that have been handled. There may be variable
  // number of them, so need to set up for next aggregate function.
  argCols = argCols[1:]
}
```
If you look at the plans that are generated, even constants are just projected input variables to the aggregates. Once we do this, we could theoretically remove support for aggregation argument constants during execution.
</blockquote></details>

I created the third commit for a problematic case.

Pay attention to below test in `/sql/opt/norm/testdata/rules/reject_nulls`.

```sql
SELECT string_agg(s, ',')
FROM (SELECT x FROM xy)
LEFT JOIN (SELECT k, s FROM a)
ON True
GROUP BY k
HAVING string_agg(s, ',')='foo'
```

I can't find a starting point for troubleshooting. Can you advise something?


<!-- Sent from Reviewable.io -->

andy-kimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44628)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball, @RaduBerinde, and @yuzefovich)

___
*[pkg/sql/opt/optbuilder/groupby.go, line 354 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44628#-M-tdpY6CFLWoJ0yvt3o:-M0SteaE5jMtrSWOZvNV:b-4ifomz) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b5940af336f3491d7a9957a2a373809937d353b/pkg/sql/opt/optbuilder/groupby.go#L354)):*
<details><summary><i>Previously, hueypark (Jaewan Park) wrote…</i></summary><blockquote>

I created the third commit for a problematic case.

Pay attention to below test in `/sql/opt/norm/testdata/rules/reject_nulls`.

```sql
SELECT string_agg(s, ',')
FROM (SELECT x FROM xy)
LEFT JOIN (SELECT k, s FROM a)
ON True
GROUP BY k
HAVING string_agg(s, ',')='foo'
```

I can't find a starting point for troubleshooting. Can you advise something?

</blockquote></details>

OK, thanks, I see the problem. It looks like we don't yet have support for `Project` in `DeriveRejectNullCols`. To add that support, add these lines to the switch:
```
	case opt.ProjectOp:
		// Pass through all null-rejection columns that the Project passes through.
		// The PushSelectIntoProject rule is able to push the IS NOT NULL filter
		// below the Project for these columns.
		rejectNullCols := DeriveRejectNullCols(in.Child(0).(memo.RelExpr))
		relProps.Rule.RejectNullCols = relProps.OutputCols.Intersection(rejectNullCols)
```
Also, add this test to `reject_nulls`:
```
# Don't reject nulls when aggregate argument is a not a Project passthrough
# column.
norm expect-not=RejectNullsGroupBy
SELECT string_agg(s || 'bar', ',')
FROM (SELECT x FROM xy)
LEFT JOIN (SELECT k, s FROM a)
ON True
GROUP BY k
HAVING string_agg(s || 'bar', ',')='foo'
----
```
If you're interested in more background information about null rejection, see the comment at the top of `reject_nulls.opt` and the comment on the `props.Relational.Rule.RejectNullColls` rule.

And here is a seminal academic paper that covers null rejection: https://www.researchgate.net/publication/220225172_Outerjoin_Simplification_and_Reordering_for_Query_Optimization


<!-- Sent from Reviewable.io -->

andy-kimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44628)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball, @RaduBerinde, and @yuzefovich)

___
*[pkg/sql/opt/optbuilder/groupby.go, line 354 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44628#-M-tdpY6CFLWoJ0yvt3o:-M0SteaE5jMtrSWOZvNV:b-4ifomz) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b5940af336f3491d7a9957a2a373809937d353b/pkg/sql/opt/optbuilder/groupby.go#L354)):*
<details><summary><i>Previously, hueypark (Jaewan Park) wrote…</i></summary><blockquote>

I created the third commit for a problematic case.

Pay attention to below test in `/sql/opt/norm/testdata/rules/reject_nulls`.

```sql
SELECT string_agg(s, ',')
FROM (SELECT x FROM xy)
LEFT JOIN (SELECT k, s FROM a)
ON True
GROUP BY k
HAVING string_agg(s, ',')='foo'
```

I can't find a starting point for troubleshooting. Can you advise something?

</blockquote></details>

OK, thanks, I see the problem. It looks like we don't yet have support for `Project` in `DeriveRejectNullCols`. To add that support, add these lines to the switch:
```
	case opt.ProjectOp:
		// Pass through all null-rejection columns that the Project passes through.
		// The PushSelectIntoProject rule is able to push the IS NOT NULL filter
		// below the Project for these columns.
		rejectNullCols := DeriveRejectNullCols(in.Child(0).(memo.RelExpr))
		relProps.Rule.RejectNullCols = relProps.OutputCols.Intersection(rejectNullCols)
```
Also, add this test to `reject_nulls`:
```
# Don't reject nulls when aggregate argument is a not a Project passthrough
# column.
norm expect-not=RejectNullsGroupBy
SELECT string_agg(s || 'bar', ',')
FROM (SELECT x FROM xy)
LEFT JOIN (SELECT k, s FROM a)
ON True
GROUP BY k
HAVING string_agg(s || 'bar', ',')='foo'
----
```
If you're interested in more background information about null rejection, see the comment at the top of `reject_nulls.opt` and the comment on the `props.Relational.Rule.RejectNullColls` rule.

And here is a seminal academic paper that covers null rejection: https://www.researchgate.net/publication/220225172_Outerjoin_Simplification_and_Reordering_for_Query_Optimization


<!-- Sent from Reviewable.io -->

hueypark reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44628)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @andy-kimball, @RaduBerinde, and @yuzefovich)

___
*[pkg/sql/opt/optbuilder/groupby.go, line 354 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/44628#-M-tdpY6CFLWoJ0yvt3o:-M0UJnL18Xg36ucjN2Cw:bxvm1c1) ([raw file](https://github.com/cockroachdb/cockroach/blob/4b5940af336f3491d7a9957a2a373809937d353b/pkg/sql/opt/optbuilder/groupby.go#L354)):*
<details><summary><i>Previously, andy-kimball (Andy Kimball) wrote…</i></summary><blockquote>

OK, thanks, I see the problem. It looks like we don't yet have support for `Project` in `DeriveRejectNullCols`. To add that support, add these lines to the switch:
```
	case opt.ProjectOp:
		// Pass through all null-rejection columns that the Project passes through.
		// The PushSelectIntoProject rule is able to push the IS NOT NULL filter
		// below the Project for these columns.
		rejectNullCols := DeriveRejectNullCols(in.Child(0).(memo.RelExpr))
		relProps.Rule.RejectNullCols = relProps.OutputCols.Intersection(rejectNullCols)
```
Also, add this test to `reject_nulls`:
```
# Don't reject nulls when aggregate argument is a not a Project passthrough
# column.
norm expect-not=RejectNullsGroupBy
SELECT string_agg(s || 'bar', ',')
FROM (SELECT x FROM xy)
LEFT JOIN (SELECT k, s FROM a)
ON True
GROUP BY k
HAVING string_agg(s || 'bar', ',')='foo'
----
```
If you're interested in more background information about null rejection, see the comment at the top of `reject_nulls.opt` and the comment on the `props.Relational.Rule.RejectNullColls` rule.

And here is a seminal academic paper that covers null rejection: https://www.researchgate.net/publication/220225172_Outerjoin_Simplification_and_Reordering_for_Query_Optimization
</blockquote></details>

Done.

Thank you for the detailed explanation.


<!-- Sent from Reviewable.io -->

andy-kimball reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>, thanks for all the work on this. Looks like we're now set up to add the other statistics-related aggregate functions.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44628#-:-M0aSCxr3RAHskYBz4GZ:b-5nzci3)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andy-kimball, @RaduBerinde, and @yuzefovich)



<!-- Sent from Reviewable.io -->

hueypark reviewed (commented):
Thank you for the review!

Could we merge this immediately? In a [sprint](https://www.sprintseoul.org/) planned tomorrow, someone could implement aggregate functions.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44628#-:-M0bcXLZAARthpMt3ilt:b1f9y3q)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andy-kimball, @RaduBerinde, and @yuzefovich)



<!-- Sent from Reviewable.io -->

RaduBerinde commented:
@hueypark there are some failing tests in the CI run (TestNormRules, TestBuilder/aggregate). Once they are addressed we can merge.

RaduBerinde commented:
If it helps, I can also push this into a temporary branch that can be used for the sprint.

hueypark commented:
> @hueypark there are some failing tests in the CI run (TestNormRules, TestBuilder/aggregate). Once they are addressed we can merge.

This is a little weird. Is there anything I missed?

> If it helps, I can also push this into a temporary branch that can be used for the sprint.

No thanks. There are still other interesting issues.

RaduBerinde commented:
You can repro with `make testrace PKG=./pkg/sql/opt/norm`

We have a bunch of assertions that only run in "race" mode. I think there's an assertion about string agg that is now incorrect (in `memo.CheckExpr`).

hueypark committed:
sql: support multi-argument aggregate functions

It is for issue #41274(sql: Support aggregate functions for statistics).

Release note: None

hueypark committed:
sql: support `corr()`

Support aggregate function `corr()`. It referred to the PostgreSQL implementation, but the implementation details follow SQL 2003. See: https://www.postgresql.org/docs/10/functions-aggregate.html

See #41274

Release note (sql change): Support aggregate function `corr()`

hueypark commented:
Thank you, I couldn't see `testrace` before you told me.

hueypark commented:
The test failed again with `TestPrimaryKeyChangeWithOperations` and something is strange. Can I get retry permission?

RaduBerinde commented:
I doubt that test has anything to do with this change. Let's try to merge.

bors r+

RaduBerinde commented:
Thanks for your contribution!
